Use Ads' currency inside Dashboard's table.

tomalec opened the pull request:
### Changes proposed in this Pull Request:

- Make `/wc/gla/ads/connection` return also currency symbol. (187d539)
	It is easier to use the code to symbol dictionary directly on the server-side. And we will need it in the majority of cases when we check the Ads currency.
- Add `useAdsCurrency` hook, use it in the Dashboard table (a4f01a5)
	Keep using currency format as specified in the shop's settings, but fetch and display Ads' account currency and symbol, to display the correct budget values.
	Solves part of https://github.com/woocommerce/google-listings-and-ads/issues/363#issuecomment-883081178https://github.com/woocommerce/google-listings-and-ads/issues/363#issuecomment-883081178


### Screenshots:

#### Woo currency settings
![image](https://user-images.githubusercontent.com/17435/127513949-13764896-c1ac-43ae-ba7e-1dfccee551bf.png)

#### Ads currency 
![image](https://user-images.githubusercontent.com/17435/126819728-19a08916-69e1-497d-b189-28cd28aa2bb0.png)

#### Dashboard table currency
![image](https://user-images.githubusercontent.com/17435/127513819-45aefe81-9adc-4a99-8115-5850a52473b6.png)



### Detailed test instructions:

1. [Set the store currency](https://gla1.test/wp-admin/admin.php?page=wc-settings) to _X_ (PLN)
2. Setup Ads account
3. Change the store currency to _Y_ (EUR)
4. Go to the Dashboard ([`/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fdashboard`](https://gla1.test/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fdashboard) )
5. You should see the "Daily budget" column using store's formatting settings, but display Ads' currency symbol.

### Changelog entry

> Fix - Render Ads Account's currency in Dashboard's table

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted. 
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Or leave the "Changelog entry" header in place without any summary if no changelog entry is needed.  
Otherwise, the title of Pull Request will be used as the changelog entry.  
-->

### Additional notes:
1. :scroll: There are some concerns about an additional sequential Google Account request as mentioned in the previous PR https://github.com/woocommerce/google-listings-and-ads/pull/901#issuecomment-887704588. We could consider removing that need in a separate PR.

tomalec committed:
Make /wc/gla/ads/connection return also currency symbol.

It is easier to use the code to symbol dictionary directly on the server-side. And we will need it in the majority of cases when we check the Ads currency.

tomalec committed:
Add useAdsCurrency hook, use it in Dashboard table

Keep using currency format as specified in the shop's settings, but fetch and display Ads' account currency and symbol, to display the correct budget values.
Solves part of https://github.com/woocommerce/google-listings-and-ads/issues/363#issuecomment-883081178

tomalec committed:
Add tests for `useAdsCurrency`.

Reaplce `isResolving` with `hasFinishedResolution` which may be more usefull, for flagging loaded state in the UI.

eason9487 commented on js/src/hooks/useAdsCurrency.test.js:
```diff
@@ -0,0 +1,124 @@
+} ) );
+
+describe( 'useAdsCurrency', () => {
+	test( 'initially should return `{ currency: Object, hasFinishedResolution: undefined }`', () => {
+		const { result } = renderHook( () => useAdsCurrency() );
+
+		// assert initial state
+		expect( result.current ).toHaveProperty(
+			'hasFinishedResolution',
+			undefined
+		);
```
❓ 
Although the `hasFinishedResolution` comes from another hook, the JSDoc of **useAdsCurrency** indicates it has only boolean type, but it actually returns `undefined` when initial. Could we make it consistent on either returned type or JSDoc for **useAdsCurrency**?

https://github.com/woocommerce/google-listings-and-ads/blob/f95a7cfd42b8de269e4a82dbd69a829c11c1a044/js/src/hooks/useAdsCurrency.js#L33

eason9487 commented on js/src/hooks/useAdsCurrency.test.js:
```diff
@@ -0,0 +1,124 @@
+			// assert initial state
+			expect( result.current.hasFinishedResolution ).toEqual( undefined );
+			// assert
+			// await act( async () => {
```
❔ 
Is this a commented out code?

eason9487 commented on js/src/hooks/useAdsCurrency.test.js:
```diff
@@ -0,0 +1,124 @@
+			await waitFor( () =>
+				expect( result.current.hasFinishedResolution ).toEqual( true )
+			);
+			// } );
```
❔ 
Same as above.

eason9487 commented on js/src/hooks/useAdsCurrency.test.js:
```diff
@@ -0,0 +1,124 @@
+				symbol: 'z\u0142',
+				status: 'connected',
+			};
+			apiFetch
+				// Mock /wp-json/wc/gla/jetpack/connected?_locale=user
+				.mockReturnValueOnce( {
+					active: 'yes',
+					owner: 'yes',
+					displayName: 'username',
+					email: 'tomalec.gla1.test@gmail.com',
+				} )
+				// Mock /wp-json/wc/gla/google/connected
+				.mockReturnValueOnce( {
+					active: 'yes',
+					email: 'gla1.test@example.com',
+				} )
+				.mockReturnValueOnce(
+					// Mock /wp-json/wc/gla/ads/connection
+					adsAccountData
+				);
```
❔ 
I'm curious why to choose to mock the deeper dependency of `apiFetch` rather than `useGoogleAdsAccount`?

tomalec replied:
To leave the fact of using `useGoogleAdsAccount` to be an implementation detail, and test observable behavior, and mock only basic/fundamental data sources.

eason9487 commented:
> There are also cases where it's more subtle and they won't notice for example CAD and USD would both show with $

Should we consider the case mentioned in https://github.com/woocommerce/google-listings-and-ads/issues/363#issuecomment-876309170 ?

tomalec committed:
Update JSDoc type for `useAdsCurrency`.

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/920#discussion_r689437447

tomalec committed:
Remove commented out code,

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/920#discussion_r689440849.

tomalec committed:
Remove unneccesary part of mocked API url from `useAdsCurrency.test`.

tomalec committed:
Merge branch 'trunk' into fix/363-ads-currency-dash-table

tomalec committed:
Render currency code instead of symbol in dashboard,

to avoid ambiguity.

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/920#issuecomment-899437561.

tomalec commented:
> Should we consider the case mentioned in #363 (comment) ?

I added a `formatAmountWithCode`. I'm not adding tests here, as I'd like to make a PR with a feature request to `@woocommerce/currency` with it.

tomalec commented:
I addressed all the comments, and the PR is ready for another round of review.

tomalec commented:
Created upstream PR at https://github.com/woocommerce/woocommerce-admin/pull/7575

eason9487 commented on js/src/dashboard/all-programs-table-card/index.js:
```diff
@@ -86,7 +93,7 @@ const AllProgramsTableCard = ( props ) => {
 				id: el.id,
 				title: el.name,
-				dailyBudget: formatAmount( el.amount ),
+				dailyBudget: formatAmountWithCode( adsCurrency, el.amount ),
```
❔ and maybe 💅 
Would it be easier to use if move the importing and calling of `formatAmountWithCode` into the `useAdsCurrency` hook?

tomalec replied:
In my mental model `useAdsCurrency`is more on data/logic level while formatting number to a string is purely presentational. That's why I'd keep it close to the component, to keep communicating, trafficking the data as `number` as far as possible.

eason9487 commented on js/src/dashboard/all-programs-table-card/format-amount-with-code.js:
```diff
@@ -0,0 +1,30 @@
+ * @param {number} number Amount to be formatted.
+ * @return {string} Amount foramtted according to given config, with the currency symbol.
+ */
+export default function formatAmountWithCode( currency, number ) {
```
💅 
Consider that there're some currency displays that need to be adjusted, this function probably would be used by them. So maybe we could move this file to a shared directory. And of course, it's good to move in future PR as well.

tomalec replied:
I was thinking about this as well, but I didn't find any current case that would use it. So, I left it closer to the place where it's used.

Probably, I'd use it in the following PRs when fixing the currency in report pages. Then, I'd move it there, because then I'd have a better judgment, what is the appropriate name and place.

eason9487 reviewed (approved):
Tested well and LGTM. Left a curious question but no need to be a blocker.
